### ISSUE TITLE
Auto-assign recommended infra type

### ISSUE THREAD

**Issue by dabreegster** [nptscot/npw#30]
After drawing a route, calculate the recommended infra type for each segment. Automatically split based on this. Maybe it'd be most clear to have this done as a user-initiated action first, instead of automatically. Have to think through how editing then works.

Also need to figure out what inputs are needed to calculate this; I think the speed (limit or observed?) of the road is the main thing

**Comment by dabreegster** [comment]
This'll require a per-road segment estimate of speed and PCU. Happy to start with tagged OSM speed limits for the first. For the second, @juanfonsecaLS1, where can I find the latest output from your model? Thanks!

**Comment by juanfonsecaLS1** [comment]
Hi Dustin, the latest results are in release [v6](https://github.com/nptscot/scottraffic/releases/tag/v6) of the scottraffic repo.

**Comment by dabreegster** [comment]
@Robinlovelace, could I ask for a code review / sanity check, please?

[get_level_of_service](https://github.com/nptscot/npw/blob/5a35b096f07bcbd5d46745fc73e9750320b73230/backend/src/level_of_service.rs#L71) implements the logic in this CbD table directly. I was originally following [osmactive](https://github.com/nptscot/osmactive/blob/349e4d607f2b93ee6ebd49bf4854f77ee9535e81/R/osmactive.R#L629), but it doesn't seem to cover all of the cases in the table. I do have some particular questions:

- How should we treat the "unknown" infra type? I'm currently assuming worst case and acting like it's mixed traffic. Now that we have automatic segmenting of routes and auto-assigning an infra type, maybe we ought to get rid of "unknown" as an option entirely. Are there cases where a user wants to indicate they'll put something there, but have not decided what?
- I'm trying to find the discussion/email about how the chosen infra types for this project map over to the categories in CbD. Is our "segregated wide" the same as "detached or remote cycle track"?
- Are "segregated narrow" and "shared footway" both "stepped or footway level cycle track"?
- When deciding the best infra type for a particular road segment, it now just depends on the inherent speed and traffic there. We try mixed traffic, cycle lane, segregated narrow, and segregated wide in order, picking the first one that makes the LoS be high, according to the table. Does this seem correct to you?

In practice you can check out these auto-recommendations by blanking out an area, then auto-importing the whole core network. There are quite a mix of optimal infra types, and they seem generally intuitive to me, but this is something to get feedback on.

Dundee:
![image](https://github.com/user-attachments/assets/1abd8948-3914-485f-9000-17666beb8497)
Edinburgh:
![image](https://github.com/user-attachments/assets/fbd9d67b-6498-4f8e-a30f-62e799c5ac1f)
Glasgow:
![image](https://github.com/user-attachments/assets/6d7d9cf0-bdac-4d83-ae56-0a88a36e34f4)

**Comment by Robinlovelace** [comment]
> * How should we treat the "unknown" infra type? I'm currently assuming worst case and acting like it's mixed traffic. Now that we have automatic segmenting of routes and auto-assigning an infra type, maybe we ought to get rid of "unknown" as an option entirely. Are there cases where a user wants to indicate they'll put something there, but have not decided what?

I think it's fine to get rid of unknown, in the current osmactive-derived infra types there are no unknowns and the extra category is not helpful IMO: 

![image](https://github.com/user-attachments/assets/72decd63-0081-427f-aae3-46fa3f6b40ad)

**Comment by Robinlovelace** [comment]
> I'm trying to find the discussion/email about how the chosen infra types for this project map over to the categories in CbD. Is our "segregated wide" the same as "detached or remote cycle track"?

No, "detached or remote" is the same as "Off road cycleway", as described here: 
![image](https://github.com/user-attachments/assets/354c2bbd-7159-48d7-8cdc-5dd025efc060)

Source: Click on the info button to the Right of "as per the **Cycling by Design guidance**" text in the screenshot above.

**Comment by Robinlovelace** [comment]
Source: https://github.com/nptscot/nptscot.github.io/blob/8a3360eaa67a1d2cfad209315ae8749235cb6da1/manual/index.md?plain=1#L292

**Comment by Robinlovelace** [comment]
> Are "segregated narrow" and "shared footway" both "stepped or footway level cycle track"?

No, segregated wide or narrow are any of the 3 segregated categories in the CbD: 

- Cycle Track at Carriageway Level
- Stepped or Footway Level Cycle Track
- Light Segregation

Source: https://www.transport.gov.scot/media/50323/cycling-by-design-update-2019-final-document-15-september-2021-1.pdf#page=56

(Just noticed a typo in our manual, we should spell that out better.)

Shared footways are ways that run parallel to a road on which you're forced to share the space with pedestrians. There was some ambiguity around shared footways that are detached (Off road in our definition), and we opted to categorise those as Off road in cases where there is clearly good provision for cycling, e.g. a high quality shared use path in the countryside.

**Comment by Robinlovelace** [comment]
> When deciding the best infra type for a particular road segment, it now just depends on the inherent speed and traffic there. We try mixed traffic, cycle lane, segregated narrow, and segregated wide in order, picking the first one that makes the LoS be high, according to the table. Does this seem correct to you?

This is a good question that we have not had to answer previously because we've been determining LoS based on infra type not vice versa. I think YES, selecting the "lowest ranked" infra type that gets us to high LoS makes sense.

**Comment by Robinlovelace** [comment]
Apologies for the slow response on this, have been working through GitHub notifications and other things on first full day back, any follow-up comments/questions welcome. One principle that was agreed with Angus was that we don't need to treat CbD as gospel. Like any guidance it has its strengths, weaknesses and subjectivities. I'm fairly confident in my answers above but don't have 100% certainty, and would be interested to learn of cases where your classification differs from `osmactive`'s. Highly conducive to auto assessment I think, so if you can share a link to a dataset with OSM ids and classification for Edinburgh, for example, that would be super useful.

**Comment by dabreegster** [comment]
Thanks for the replies. I've removed the Unknown category, because I can't come up with a reason it could be useful now.

If I'm interpreting your replies correctly, then we don't have a clear mapping from our infrastructure types to the CbD table, and thus there's not a clear way for us to derive LoS. I'm making some choices in 
https://github.com/nptscot/npw/blob/5a35b096f07bcbd5d46745fc73e9750320b73230/backend/src/level_of_service.rs#L111 that I hope are reasonable (and if not, happy to change them). One of them is treating Segregated wide as high LoS automatically, because it's the best choice available for roads, and the user needs some way to achieve high LoS there.

There's already a way to compare NPW's assigned LoS with what's in osmactive. In NPW:

1) Start with an unedited map
2) Import existing routes from OSM
3) Enable the LoS layer, and toggle between "Show original data" and "Show current derived data", looking for diffs

One example diff I noticed is https://www.openstreetmap.org/way/33465852. Following this NPW procedure, it's medium, but it's high in osmactive. There are some existing painted lanes there that are detected:
![image](https://github.com/user-attachments/assets/1c27fba0-09e8-4f36-96a7-c178c41db14a)
But we made the decision to not import painted lanes from OSM (or at least, to give the user an option there and not do it by default), even when painted lanes make the LoS go from medium to high. So, I'd like to revisit that decision too. If I manually draw a route covering that section, it gets auto-assigned as a painted cycle lane, and becomes high LoS.

**Comment by Robinlovelace** [comment]
> we don't have a clear mapping from our infrastructure types to the CbD 

I think it's clear but not 1-2-1 mapping. 

Basically:

- Detached -> Offroad in NPT
- Protected (all 3 categories in CbD) -> Segregated (wide or narrow depending on width) in NPT
- Protected but shared use (not a category in CbD) -> Shared footway in NPT
- Mixed traffic -> ignored in NPT
- Cycle Lane -> Painted lane

Does that look right? May have missed something, would be happy to hear clarifications and details regarding which aspects of the lookup are unclear.

**Comment by Robinlovelace** [comment]
> But we made the decision to not import painted lanes from OSM (or at least, to give the user an option there and not do it by default), even when painted lanes make the LoS go from medium to high. So, I'd like to revisit that decision too.

Happy to revisit that but not my decision.

**Comment by dabreegster** [comment]
Because the "cycle track at carriageway level", "stepped cycle track", and "light segregation" CbD categories don't have a clear mapping to "segregated wide" and "segregated narrow", it's unclear what LoS logic I should apply in https://github.com/nptscot/npw/blob/5a35b096f07bcbd5d46745fc73e9750320b73230/backend/src/level_of_service.rs#L111. I'll just review this function directly with Angus next meeting.

> Mixed traffic -> ignored in NPT

We do have mixed traffic in NPW, and in fact it gets auto-recommended in some places, based on low speed and traffic.

**Comment by Robinlovelace** [comment]
Yes, we lump together "cycle track at carriageway level", "stepped cycle track", and "light segregation". That was a design decision and a good one I think. The differences in LoS only affect 4 cells in the LoS categorisation table, as shown below, right?

![image](https://github.com/user-attachments/assets/dbd78bc9-3142-4cce-a4e0-12dfdf91da4f)

In each case I think it's safe to assume the worst (light segregation) LoS.

What do you think?

**Comment by Robinlovelace** [comment]
> In each case I think it's safe to assume the worst (light segregation) LoS.

Except maybe for the fastest roads where Segregated (wide) -> low LoS rather than X.

**Comment by dabreegster** [comment]
If there's no difference in LoS for segregated wide vs narrow, why do we distinguish the two? When would the user choose one over the other -- is LoS the main thing we judge everything by, or is it something else? Because for reachability questions (see https://docs.google.com/document/d/1aZ0urdYrUyG9c41MdPWnC5K70-y-PuzP_rHldPaaTio/edit?usp=sharing), it's all based on LoS.

And if we take the worst / light segregation LoS case for our segregated wide & narrow types, then there's no way for a user to achieve high LoS on high volume, high speed roads. Do we want to force them there to draw entirely new off-road segments?

**Comment by Robinlovelace** [comment]
> If there's no difference in LoS for segregated wide vs narrow, why do we distinguish the two? 

There should be a difference: cycleways must be wide enough according to CbD guidance, right?

**Comment by Robinlovelace** [comment]
I don't see the benefit of allowing the user to choose to build a narrow (non CbD compliant) cycle track, I assumed all new cycle tracks would be wide.

**Comment by Robinlovelace** [comment]
>  if we take the worst / light segregation LoS case for our segregated wide & narrow types, then there's no way for a user to achieve high LoS on high volume, high speed roads.

See my response here: https://github.com/nptscot/npw/issues/30#issuecomment-2569090114

However, according to the table, the best that any kind of segregated infrastructure can get is low LoS for 80 kph roads, right, so high LoS is not possible with alongside infrastructure on those roads, right?

**Comment by Robinlovelace** [comment]
More broadly and taking a bit of a step back, I'm not against the idea of allowing users to choose a wider range of infrastructure types in NPW than the infrastructure types we're using to visualise existing cycle infrastructure in NPT.

**Comment by Robinlovelace** [comment]
However, my sense is that "start simple and then add complexity when needed/beneficial" is a good approach that we can apply to cycle infrastructure categories. Open to other ideas and happy to bounce more ideas and think about pros and cons of different options also, and keen to hear what others think, what's your take on how to handle this Dustin?

**Comment by dabreegster** [comment]
I think we need to back up and have a wider discussion about this. At the end of the day, NPW needs a set of infrastructure types and a way to judge LoS based on infra type, traffic, and speed. Maybe https://github.com/nptscot/npw/blob/5a35b096f07bcbd5d46745fc73e9750320b73230/backend/src/level_of_service.rs#L111 is already adequate for that.

The other strawman idea to drive the point home is to literally have 3 categories -- high, medium, low LoS. Based on traffic and volume, many roads will be medium or high without any infrastructure. The user can just specify they want to upgrade something, and ignore details about what specifically goes there.

To be clear, I'm not proposing this strawman, but if the only point of the infra type is to derive LoS, then something like this is the simplest. But I think we probably do care about infra type beyond the LoS it implies; I'm just trying to figure out exactly how, so it can be surfaced appropriately in NPW. Perhaps due to the coherence metric.

**Comment by dabreegster** [comment]
There've been bigger changes here. Cleaning up old issues to regroup and make fresh ones.
